### PR TITLE
updates with python upgrade 

### PR DESCRIPTION
**Pull request by samathad2023**
## Description

This PR aims to upgrade project to python version 3.12.2.This PR encompasses changes in dependencies, configurations and code modifications to ensure compatibility with Python 3.12.2 .

This should include:

 install Python 3.12 via pyenv `pyenv install 3.12`
 Create the new virtual environment against 3.12 
`cd notifications-utils
pyenv virtualenv 3.12 notify-admin-upgrade
pyenv local notify-admin-upgrade`
 Update the project to use the new virtual env that is 3.12-based (in the project root directory, `pyenv local notify-admin-upgrade
`)
 Run `make bootstrap` to install the dependencies in the new virtual env

### CODE REVIEW

File: tests/app/main/views/service_settings/test_service_settings.py [hunk 1]
```diff
@@ -2306,10 +2306,10 @@ def test_resume_service_after_confirm(
         ),
     )
 
-    assert mock_api.called_once_with(
+    mock_api.assert_called_once_with(
         "/service/{}/resume".format(SERVICE_ONE_ID), data=None
```
Comment (xlorepdarkhelm) [inline tests/app/main/views/service_settings/test_service_settings.py:2310]: I've been trying to update all the "".format() over to f strings when possible, for clarity and simplification, as well as idiomatic python purposes, could this line be changed to that?

`f"/service/{SERVICE_ONE_ID}/resume", data=None`

File: tests/app/main/views/test_find_users.py [hunk 2]
```diff
@@ -23,7 +23,7 @@ def test_find_users_by_email_displays_users_found(
     mocker.patch(
         "app.user_api_client.find_users_by_full_or_partial_email",
         return_value={"data": [user_json()]},
-        autospec=True,
+        # autospec=True, # not need with 3.12
```
Comment (xlorepdarkhelm) [inline tests/app/main/views/test_find_users.py:26]: I would recommend removing the line as opposed to commenting it out.

File: tests/app/main/views/test_find_users.py [hunk 3]
```diff
@@ -57,7 +57,7 @@ def test_find_users_by_email_displays_multiple_users(
         return_value={
             "data": [user_json(name="Apple Jack"), user_json(name="Apple Bloom")]
         },
-        autospec=True,
+        # autospec=True, # not need with 3.12
```
Comment (xlorepdarkhelm) [inline tests/app/main/views/test_find_users.py:60]: All of these commented out lines should be removed.

File: tests/app/models/test_base_model.py [hunk 4]
```diff
@@ -21,7 +21,10 @@ def foo(self):
     with pytest.raises(AttributeError) as e:
         Custom({"foo": "NOPE"})
 
-    assert str(e.value) == "can't set attribute"
+    assert (
+        str(e.value)
+        == "property 'foo' of 'test_raises_when_overriding_custom_properties.<locals>.Custom' object has no setter"
+    )
```
Comment (xlorepdarkhelm) [inline tests/app/models/test_base_model.py:27]: Eww, the bad side of the new descriptive errors for Python shows its ugly head.

**Review by xlorepdarkhelm — CHANGES_REQUESTED** [review]
Just a few changes. Nothing major. The conversion to 3.12 doesn't seem to drastic.

File: tests/app/main/views/service_settings/test_service_settings.py (hunk 1 above)
Comment (samathad2023) [reply tests/app/main/views/service_settings/test_service_settings.py:2310]: This should work with f"/service/{SERVICE_ONE_ID}/resume" . I will update.

File: tests/app/models/test_base_model.py (hunk 4 above)
Comment (xlorepdarkhelm) [reply tests/app/models/test_base_model.py:27]: (nothing to change here, just disturbing how much extra text needs to be verified now)

Comment (samathad2023) [reply tests/app/models/test_base_model.py:27]: yes, utils have more asserts like this (more descriptive messaging in 3.12)

**Review by ccostino — APPROVED** [review]
Thank you, @samathad2023!

**Comment by ccostino** [comment]
Due to timing/availability of folks and the fact that we now have the Python updates merged for both the API and utils repos, I'm dismissing @xlorepdarkhelm's previous review so I can merge this PR.

I double checked and verified that @samathad2023 had addressed all of the requested changes and incorporated all of the feedback @xlorepdarkhelm requested.  Thank you for the thorough review, @xlorepdarkhelm, and thank you @samathad2023 for making the necessary updates/improvements! 🎉

**Pull request merged**